### PR TITLE
Feat: Typechain + Ethers instead of Web3js

### PR DESCRIPTION
**Pull request by josipbagaric**
- [x] Use [TypeChain](https://github.com/ethereum-ts/TypeChain) to auto-generate contract wrappers for Typescript
- [x] Migrated `@energyweb/issuer` to use [ethers](https://github.com/ethers-io/ethers.js/) instead of Web3.js, for easier tracking of contract events and fixing some of the Web3js bugs
- [x] **Only show energy volumes for the active user**: Certificate.energy now returns "publicVolume", "privateVolume" and "claimedVolume" https://github.com/energywebfoundation/origin/pull/811/commits/39cf708d8b02e8859b6ec48babbadfcffc53c32b
- [x] Adjust `@energyweb/origin-ui-core` to use ethers instead of Web3js
- [x] Remove references to `issuedCertificateId` in Issuer.sol https://github.com/energywebfoundation/origin/pull/811/commits/d9ce8c2daae6ccde9c7515db96ef76c61456a80e
- [x] Improved revokation workflow https://github.com/energywebfoundation/origin/pull/811/commits/1e462d617b4ba92ea4fea7874dd7db4b17d2bdc4
- [x] Support for BigNumber in precise proofs (https://github.com/energywebfoundation/precise-proofs/pull/8)
- [x] Store energy values (up to Solidity uint256) as `string` in PostgreSQL https://github.com/energywebfoundation/origin/pull/811/commits/2a284a9ff3b362dfa41516995b1cadd4c5651194

### CODE REVIEW

File: packages/issuer/src/blockchain-facade/CertificateUtils.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    // TO-DO: replace with proper claim data
+    const claimData = certificates.map(() => randomBytes(32));
+    const data = randomBytes(32);
+
+    const { activeUser } = configuration.blockchainProperties as Configuration.BlockchainProperties<
+        Registry,
+        Issuer
+    >;
+
+    const { registry } = configuration.blockchainProperties as Configuration.BlockchainProperties<
```
Comment (e00dan) [inline packages/issuer/src/blockchain-facade/CertificateUtils.ts]: Why not in the same line as activeUser assignment?

File: packages/origin-ui-core/webpack/base.config.js [hunk 2]
```diff
@@ -71,5 +71,9 @@ module.exports = {
 
     node: {
         fs: 'empty'
-    }
+    },
+
+    externals:[{
+        xmlhttprequest: '{XMLHttpRequest:XMLHttpRequest}'
```
Comment (e00dan) [inline packages/origin-ui-core/webpack/base.config.js:77]: Why do we need that? : /

**Review by e00dan — APPROVED** [review]

File: packages/origin-ui-core/webpack/base.config.js (hunk 2 above)
Comment (josipbagaric) [reply packages/origin-ui-core/webpack/base.config.js:77]: TBH can't remember. There was some error and this fixed it. Will check again.

File: packages/issuer/contracts/ERC1888/IERC1888.sol [hunk 3]
```diff
@@ -9,7 +9,7 @@ interface ERC1888 {
         bytes data;
     }
 
-    event IssuanceSingle(address indexed _issuer, int256 indexed _topic, uint256 _id);
+    event IssuanceSingle(address indexed _issuer, int256 indexed _topic, uint256 indexed _id);
```
Comment (kosecki123) [inline packages/issuer/contracts/ERC1888/IERC1888.sol]: This is violating the ERC1888

File: packages/issuer/contracts/Issuer.sol [hunk 4]
```diff
@@ -7,22 +7,21 @@ import "@openzeppelin/contracts-ethereum-package/contracts/ownership/Ownable.sol
 import "./Registry.sol";
 
 contract Issuer is Initializable, Ownable {
-    event NewCertificationRequest(address indexed _owner, uint256 indexed _id);
+    event NewCertificationRequest(address indexed _owner, uint256 indexed _id, string indexed _deviceId);
```
Comment (kosecki123) [inline packages/issuer/contracts/Issuer.sol]: Ideally we would keep the naming scheme such as: 

"All events should be represented as verbs in the past tense such as CustomerRelocated, CargoShipped, etc."

File: packages/issuer/contracts/Issuer.sol [hunk 5]
```diff
@@ -7,22 +7,21 @@ import "@openzeppelin/contracts-ethereum-package/contracts/ownership/Ownable.sol
...
     event ApprovedCertificationRequest(address indexed _owner, uint256 indexed _id, uint256 indexed _certificateId);
+    event RevokedCertificationRequest(address indexed _owner, uint256 indexed _id);
 
 	event CommitmentUpdated(address indexed _owner, uint256 indexed _id, bytes32 _commitment);
 	event MigrateToPublicRequested(address indexed _owner, uint256 indexed _id);
 	event PrivateTransferRequested(address indexed _owner, uint256 indexed _certificateId);
 	event CertificateMigratedToPublic(uint256 indexed _certificateId, address indexed _owner, uint256 indexed _amount);
+    event RevokedCertificate(uint256 indexed _certificateId);
```
Comment (kosecki123) [inline packages/issuer/contracts/Issuer.sol]: Ideally we would keep the naming scheme such as: 

"All events should be represented as verbs in the past tense such as CustomerRelocated, CargoShipped, etc."

File: packages/issuer/contracts/Issuer.sol [hunk 6]
```diff
@@ -130,34 +110,47 @@ contract Issuer is Initializable, Ownable {
...
+            && request.approved
+            && !request.revoked
+            && !revokedCertificates[certificateId];
     }
 
-    function revokeRequest(uint256 _requestId) public onlyOwner {
+    function revokeRequest(uint256 _requestId) external {
         CertificationRequest storage request = certificationRequests[_requestId];
+
+        require(msg.sender == request.owner, "revokeRequest(): Only the request creator can revoke the certificate.");
```
Comment (kosecki123) [inline packages/issuer/contracts/Issuer.sol]: Comment: ...revoke the request

File: packages/migrations/src/schema/create_tables.sql [hunk 7]
```diff
@@ -76,7 +76,7 @@ CREATE TABLE public.certification_request (
     "createdAt" timestamp with time zone DEFAULT now() NOT NULL,
     "updatedAt" timestamp with time zone DEFAULT now() NOT NULL,
     id integer NOT NULL,
-    energy integer NOT NULL,
+    energy character varying NOT NULL,
```
Comment (kosecki123) [inline packages/migrations/src/schema/create_tables.sql]: This should go to the migrations stuff. Which is still WIP #851

File: packages/origin-backend/src/utils/positiveBNStringValidator.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { ValidatorConstraint, ValidatorConstraintInterface } from 'class-validator';
+import { bigNumberify, BigNumber } from 'ethers/utils';
+
+@ValidatorConstraint()
+export class PositiveBNStringValidator implements ValidatorConstraintInterface {
```
Comment (kosecki123) [inline packages/origin-backend/src/utils/positiveBNStringValidator.ts:5]: This should be a part of utils or other common package to exchange and backend

**Review by kosecki123 — CHANGES_REQUESTED** [review]
2 blockers:

- ERC1888 interface change
- Database schema change in the create script - this will break the incoming changes in #851

File: packages/migrations/src/schema/create_tables.sql (hunk 7 above)
Comment (josipbagaric) [reply packages/migrations/src/schema/create_tables.sql]: It's needed for this to work. Should I wait for the WIP PR to be done?

File: packages/origin-backend/src/utils/positiveBNStringValidator.ts (hunk 8 above)
Comment (josipbagaric) [reply packages/origin-backend/src/utils/positiveBNStringValidator.ts:5]: I agree, though this one is not the same as the one in the exchange.

File: packages/migrations/src/schema/create_tables.sql (hunk 7 above)
Comment (kosecki123) [reply packages/migrations/src/schema/create_tables.sql]: Yes, I think so, let's keep this one open, the #851 should be done soon.

**Review by kosecki123 — APPROVED** [review]
LGTM, let's wait for #851 now.

**Pull request merged**